apps/trackers/service.py

```diff
@@ -0,0 +1,105 @@
+from jira.exceptions import JIRAError
+from rest_framework.response import Response
+
+from collectors.jiraffe.core import JiraQuerier
+from osidb.models import Affect, PsUpdateStream, Tracker
+
+from .bts_tracker import JiraTracker
+from .constants import JIRA_SERVER
+
+logger = logging.getLogger(__name__)
+
+
+class JiraTrackerQuerier(JiraQuerier):
+    """
+    Jira query handler for tracker management.
+    This class encapsulates calls for Jira doing validations
+    and it methods return data requested with HTTP status code
+    """
+
+    def __init__(self, token) -> None:
+        """
+        Instantiate a new JiraTrackerQuerier object.
+
+        Keyword arguments:
+        token -- user token used in every request to Jira
+        """
+        self._jira_server = JIRA_SERVER
+        self._jira_token = token
+
+    def get_bts_tracker(self, bts_id):
+        """get Jira tracker given its string key or integer id"""
+        try:
+            return Response(data=self.jira_conn.issue(bts_id).raw, status=200)
+        except JIRAError as e:
+            return Response(data=e.response.json(), status=e.status_code)
+
+    def create_bts_affect_tracker(
+        self, affect: Affect, tracked_streams: list[PsUpdateStream]
+    ):
+        """
+        Generates a tracker in BTS given an Affect.
+        If multiples streams is given creates one tracker per stream.
+        """
+        if affect.trackers.exists():
+            return Response(data="Affect already have trackers created.", status=409)
+
+        # creates a list of trackers to be generated in BTS
+        bts_trackers = []
+        if not tracked_streams:
+            bts_trackers.append(
+                JiraTracker(flaw=affect.flaw, affect=affect, stream=None)
+            )
+        else:
+            for stream in tracked_streams:
+                bts_trackers.append(
+                    JiraTracker(
+                        flaw=affect.flaw,
+                        affect=affect,
+                        stream=stream,
+                    )
+                )
+        # submits a list of trackers/Jira issues to be created in bulk
+        try:
+            bulk_issues = []
+            for bts_tracker in bts_trackers:
+                bulk_issues.append(bts_tracker.generate_bts_object()["fields"])
+
+            response = self.jira_conn.create_issues(bulk_issues, prefetch=True)
+            issues = response
+        except JIRAError as e:
+            return Response(data=e.response.json(), status=e.status_code)
+
+        # creates in OSIDB the equivalent entity for the created BTS trackers
+        issues = [created["issue"] for created in response]
+        try:
+            with transaction.atomic():
+                i = 0
+                while i < len(response):
+                    issue = issues[i].raw
+                    bts_tracker = bts_trackers[i]
+                    stream = "" if not bts_tracker._stream else bts_tracker._stream.name
```

Comment:
We should not permit creating new trackers without the stream. There can be some historical or invalid missing it but the newly created should be all valid.

Reply:
Makes sense, enforced the stream.